File: cpp/Single_Number_II.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

class Solution {
public:
  int singleNumber(vector<int>& nums) {
    int thn=-1, thnp1=0, thnp2=0;
    for(int &num : nums){
      int thn_mask = thn & num;
      int thnp1_mask = thnp1 & num;
      int thnp2_mask = thnp2 & num;

      thnp1 |= thn_mask;
      thn &= ~(thn_mask);

      thnp2 |= thnp1_mask;
      thnp1 &= ~(thnp1_mask);

      thn |= thnp2_mask;
      thnp2 &= ~(thnp2_mask);
    }
    return thnp1;
  }
};

int main(){
  vector<int> vec;
  string input, inp;
  getline(cin, input);
  stringstream input_stream(input);
  while (getline(input_stream, inp, ',')){
    vec.push_back(stoi(inp));
  }
  Solution sol;
  cout <<sol.singleNumber(vec) <<endl;
  return 0;
}

File: ruby/single_number_II.rb
# @param {Integer[]} nums
# @return {Integer}
def single_number(nums)
   tn, tnp1, tnp2 = 4294967295, 0, 0
   nums.each do |num|
    tn = tn & ~(num & tn)
    tnp1 = tnp1 | (num & tn)

    tnp1 = tnp1 & ~(num & tnp1)
    tnp2 = tnp2 | (num & tnp1)

    tnp2 = tnp2 & ~(num & tnp2)
    tn = tn | (num & tnp2)
   end
   puts 'tn', tn
   puts 'tn + 1', tnp1
   puts 'tn + 2', tnp2
end

single_number([0,1,0,1,0,1,99])
